feat(fieldset): add Fieldset and ValidationMessage

thallada opened the pull request:
Also, refactors the invalid-feedback div out of asInput into its own component.

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+
+## API
+
+### `dangerIconDescription` (string or element; optional)
```
"danger" seems implementation (aka `Bootstrap`)-specific. This feels like it's some kind've "invalidInputDescription" or something #NamingIsHard

jaebradley replied:
Why would somebody use an element here?

thallada replied:
studio-frontend uses an element here because we wrap every translated string in a span that indicates the language of the string with the `lang` attribute.

It's called `dangerIconDescription` because it describes the exclamation icon that only appears when the danger theme is applied.

I'm really just copying what `asInput` is doing. I personally don't like the whole theme thing because I don't see us ever needing an input/fieldset styled with anything other than red or no-red, and it's not clear what happens if you specify multiple themes in the array.

I could completely redo the API for Fieldset, but then it would be inconsistent with asInput. We cannot also redo the API for asInput without making a breaking change release. I opted to make the least changes and inconsistencies and just copied the API for asInput. If we have a clear plan for what the API should be, then I can adopt that here and then we can update asInput to it in a separate PR.

fysheets replied:
Want to note that it could be styled with yellow warning or any of the other bootstrap coloring variant/themes as well even though we're not doing that for SFE right now.

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+`dangerIconDescription` can be used to provide a screen-reader description of the "exclamation" icon used in the validation message displayed when `isValid` is false and `themes` includes "danger". The default is an empty string.
+
+### `children` (string, element, or arrays of strings or elements; optional)
+`children` specifies the elements that the Fieldset will wrap. The default is an empty string.
```
Is there a reason to explicitly point out the `children` prop vs. the implicit use of the `children` prop when using `Fieldset` as a wrapper.

I could see how somebody might get confused with this `prop` and do something like

```javascript
<FieldSet children={<noblah />)>
  <blah />
</FieldSet>
```

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+`className` specifies Bootstrap class names to apply to the input component. The default is an empty array.
+
+### `id` (string; optional)
+`id` specifies a unique identifier for the fieldset.  It will be used to create an id for the ValidationMessage that is referenced by an `aria-describedby` on the `fieldset`. The default is an auto-generated id of the form `fieldsetN` where N is an auto-incremented integer.
```
`fieldsetN` is typo?

jaebradley replied:
lol no it's not, I am

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+`legend` specifies the label to be used for the overarching fieldset. This can be a string or element type. It appears at the top of the fieldset, inline with the border.
+
+### `themes` (array of strings; optional)
+`themes` specifies the themes to apply to the input (e.g. "danger"). The default is an empty array.
```
Maybe clarify these are one of [the eight `Bootstrap` contextual variations](https://getbootstrap.com/docs/4.0/components/badge/#contextual-variations)?

thallada replied:
Isn't that docs page specific to badges though? I don't see any docs pages in Bootstrap about the different conceptual themes they have (danger, warning, info).

fysheets replied:
https://v4-alpha.getbootstrap.com/utilities/colors/

thallada replied:
Thanks Farhannah, I found the v4.0 docs for that and linked it in the docs here: https://getbootstrap.com/docs/4.0/utilities/colors/

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+### `themes` (array of strings; optional)
+`themes` specifies the themes to apply to the input (e.g. "danger"). The default is an empty array.
+
+### `validationMessage` (string or element; optional)
```
If it's only displayed when `isValid` is `false`, maybe this is more of an `invalidMessage`?

thallada replied:
Yes, that's probably a little clearer, but this would be another change that makes the Fieldset API inconsistent with asInput.

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+### `className` (string array; optional)
+`className` specifies Bootstrap class names to apply to the input component. The default is an empty array.
+
+### `id` (string; optional)
```
If somebody wants to use an `id` of `1` will this result in a `PropType` error? Or will `1` get stringified in-time?

thallada replied:
I think that would result in a proptype error, but I can verify that.

thallada replied:
![screenshot-20180420143153-969x109](https://user-images.githubusercontent.com/1505923/39067696-9e463a38-44a7-11e8-9213-c00b9d9cd091.png)

Yup, PropType error.

jaebradley replied:
Why wouldn't an `id` value of `1` be valid?

thallada replied:
I think supporting both would unnecessarily complicate things because we later do string operations on the id because it is assumed to be a string. Also, I feel like an `id` attribute that is just a number is not very descriptive.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,79 @@
+  constructor(props) {
+    super(props);
+
+    const id = this.props.id ? this.props.id : newId('fieldset');
```
[`airbnb/javascript#15.7`](https://github.com/airbnb/javascript#comparison--unneeded-ternary)

Consider

```javascript
const id = this.props.id || newId('fieldset');
```

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  render() {
+    return (
```
consider destructuring `this.props` and `this.state`

jaebradley commented on src/ValidationMessage/index.jsx:
```diff
@@ -0,0 +1,85 @@
+
```
Extra line

jaebradley commented on src/ValidationMessage/index.jsx:
```diff
@@ -0,0 +1,85 @@
+const defaultProps = {
+  dangerIconDescription: '',
+  className: [],
+  errorId: undefined,
```
`''` vs. `undefined`?

jaebradley commented on src/ValidationMessage/index.jsx:
```diff
@@ -0,0 +1,85 @@
+  constructor(props) {
+    super(props);
+
+    const id = this.props.errorId ? this.props.errorId : `error-${newId('validationMessage')}`;
```
There's an implicit dependency between the `error-{id}` format here and the format that's defined in the initial state declaration in `Fieldset` (aka, if we change something in one place we probably need to change it in the other place) - this seems kind've rigid?

jaebradley replied:
Nit: if you call this variable `errorId` then you can just do 

```javascript
this.state = { errorId };
```

on the next line

thallada replied:
I refactored this so that `ValidationMessage` has a required `id` prop and no state. `Fieldset` and `asInput` now create and pass the `error-{id}` themselves. I think that makes the most sense since `Fieldset` and `asInput` will have to use that same id in their `aria-describedby`.

jaebradley commented on src/ValidationMessage/index.jsx:
```diff
@@ -0,0 +1,85 @@
+            </span>
+          </span>
+          ),
+          <span key="1">
```
Since `React` uses `key`s to determine if something has changed if the `this.props.validationMessage` changes will this not be reflected in subsequent renders?

thallada replied:
I'm going to switch this code over to wrap the list of elements in `<React.Fragment>`s instead to avoid this `key`s issue.

jaebradley replied:
That will make this component incompatible with anything before `React` 16 right?

Is this something we need to be concerned about? (Probably not)

thallada replied:
I'm not aware of anything that's using Paragon and not on React 16 so I don't think it is something we should be concerned about.

thallada commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,29 @@
+### `children` (string, element, or arrays of strings or elements; optional)
+`children` specifies the elements that the Fieldset will wrap. The default is an empty string.
+
+### `className` (string array; optional)
```
I agree that this is confusing. However, I think we should keep this as `className` but make it `string` type instead. That way it is consistent with the React API and also familiar to React programmers. If users want to specify multiple classes, then they can use the `classNames` function in the value for `className` just like any normal base React component.

But, if I make that change here, I'd also want to make sure we make the change for every `className` prop in Paragon (which would be a breaking change).

thallada replied:
I talked with Jae in-person. I think we are okay with making changes like this one just in Fieldset for now. Eventually, we can make the breaking changes in the rest of Paragon to resolve the inconsistencies.

fysheets commented:
One more overarching comment -- with this refactor to pull validation out, can we also refactor inputs to use the `variant` naming instead of `theme` (that whole discussion we had with this and implemented with the modal) so that we can think of it as different types of inputs/fieldsets/components/etc instead of it being themed like we theme other edX related items? Or would that end up being a breaking change as well?

fysheets commented on src/asInput/index.jsx:
```diff
@@ -37,7 +37,7 @@ export const inputProps = {
   onChange: () => {},
   onBlur: () => {},
-  id: newId('asInput'),
+  id: undefined,
```
Mostly curious about why this change was made.

thallada replied:
The problem with this is that `defaultProps` is only defined once per import of this module. So every `asInput` created without an id specified would get this same id. The code in the constructor, which is called on every new `asInput` component, should call `newId` if this is `undefined` now.

jaebradley replied:
I was thinking about this recently - I think we should switch over to something like https://github.com/dylang/shortid ?

fysheets commented on src/asInput/index.jsx:
```diff
@@ -109,43 +109,14 @@ const asInput = (WrappedComponent, inputType = undefined, labelFirst = true) =>
       desc.error = (
-        <div
-          className={classNames(
-            styles['form-control-feedback'],
```
`form-control-feedback` is a bootstrap class that adds a bit more formatting for us, curious on why it was removed? https://v4-alpha.getbootstrap.com/components/forms/ (search for the class name)

fysheets replied:
Looking more into those docs, we weren't using it correctly is my guess on the answer -- so new question, should we be?

thallada replied:
I think `form-control-feedback` was removed in Bootstrap v4.0. It also looks like Bootstrap only supports green or red validation themes now (`is-valid` and `is-invalid` classes): https://getbootstrap.com/docs/4.0/components/forms/#server-side

fysheets commented on src/asInput/asInput.scss:
```diff
@@ -3,8 +3,8 @@
 @import '~bootstrap/scss/input-group';
 
-.fa-icon-spacing {
-    padding: 0px 5px 0px 0px;
```
Curious about this CSS change as well, as it doesn't look to me to be a simple swap.

fysheets replied:
Nevermind, I started from the bottom up and made it to the new scss file that this was moved to :)

thallada commented:
@jaebradley and @fysheets, I think I have addressed all of your comments. `Fieldset` and `ValidationMessage` now use `Variant` and have less confusing prop names.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+  isValid: PropTypes.bool,
+  legend: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
+  invalidMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
+  variant: PropTypes.shape({
```
Are there plans to add more `variant` options in the future? A `shape` with one field feels a little over-structured...

thallada replied:
I didn't create Variant, so I'm not 100% sure what the purpose of it is. It looks like it was designed so that there would be different Variant "types". The only one right now is `status`: `INFO`, `WARNING`, `DANGER`, `SUCCESS`. But, I don't know what other Variant types are possible and don't want to collapse this to a string if we'd want to add more variants later.

jaebradley replied:
What do you mean by

> don't want to collapse this to a string if we'd want to add more variants later

Couldn't the props just be `buttonVariant`, `inputVariant`, `anotherInputVariant`, etc. (assuming we want multiple variants).

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+const inputProps = {
+  variantIconDescription: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
+  children: PropTypes.node,
+  className: PropTypes.arrayOf(PropTypes.string),
```
`classNames`?

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+  className: PropTypes.arrayOf(PropTypes.string),
+  id: PropTypes.string,
+  isValid: PropTypes.bool,
+  legend: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
```
Nit: do you mind separating the required and optional `props`? I know I'm OCD, but it makes it a little easier to see at a glance.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+    super(props);
+
+    const id = props.id || newId('fieldset');
+    this.state = {
```
style nit - consider doing

```javascript
this.state = { id };
```

and save two lines.

or

```javascript
this.state = { id: props.id || newId('fieldset') };
```

and save three.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+  componentWillReceiveProps(nextProps) {
+    // TODO: use getDerivedStateFromProps when upgraded to React 16.3+
+    if (nextProps.id !== this.props.id) {
+      if (nextProps.id) {
```
Can this be simplified to

```javascript
this.setState({ id: nextProps.id || newId('fieldset') });
```

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  getVariantFieldsetClassName() {
+    const { variant } = this.props;
```
Can this be simplified to

```javascript
if (this.props.variant.status === Variant.status.INFO) {
  return styles['is-invalid-nodanger'];
}
```

thallada replied:
That is simpler, but the current structure makes it really easy to add another variant.

jaebradley replied:
Fair - I guess shouldn't that be something the implementer that wants to add more variants should solve? Especially since it's unclear if more variants will be needed (something something pre-mature optimization)?

Either way, not a big deal.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+            this.getVariantFieldsetClassName(),
+            className,
+          )}
+          aria-describedby={`error-${this.state.id}`}
```
move `error-${this.state.id}` to variable?

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  getVariantFieldsetClassName() {
```
`getVariantClassName` vs. `getVariantFieldsetClassName` - isn't the `Fieldset` implied?

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,41 @@
+* Displays `invalidMessage` in red.
+* Displays an "exclamation" icon next to `invalidMessage`.
+
+### `variantIconDescription` (string or element; optional)
```
If `variant` is going to be an object, wouldn't it make sense for it's icon description to live within that object?

thallada replied:
I think `variant` is intended just to specify one theme out of an enum. If we start putting display strings in it then I'm afraid we'll start dumping any theme-related data into `variant`.

jaebradley commented on src/Fieldset/Fieldset.stories.jsx:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  handleSubmit(event) {
+    if (this.firstInputRef.value.length === 0 && this.secondInputRef.value.length === 0) {
```
can this just be simplified to

```javascript
this.setState({ isValid: this.firstInputRef.value.length !== 0 || this.secondInputRef.value.length !== 0 });
```

jaebradley replied:
Whoops, forgot this was in a story. Carry on.

jaebradley commented on src/Fieldset/index.jsx:
```diff
@@ -0,0 +1,105 @@
+  isValid: PropTypes.bool,
+  invalidMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
+  variant: PropTypes.shape({
+    status: PropTypes.string,
```
Should this be `PropTypes.oneOf(Variant.status)`? We probably want to have `prop` warnings if somebody puts in `foobar` or something?

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,41 @@
+### `children` (node; optional)
+`children` specifies the elements that the Fieldset will wrap. The default is null.
+
+### `className` (string array; optional)
```
Should just be `string` not `string array`

jaebradley commented on src/Fieldset/README.md:
```diff
@@ -0,0 +1,41 @@
+`legend` specifies the label to be used for the overarching fieldset. This can be a string or element type. It appears at the top of the fieldset, inline with the border.
+
+### `variant` (object with a string `status` field; optional)
+`variants` specifies which type of fieldset to display. These are the Variants currently supported:
```
`variant` vs. `variants`
